### ISSUE TITLE
Docker compose issues, accessing newly added packages.

### ISSUE THREAD

**Issue by AbhinavReddy-Dev** [ChicoState/PantryNode#139]
Issues in frontend and backend accessing packages in node_modules. #125

**Comment by AbhinavReddy-Dev** [comment]
It started working for me once I removed the previous images and built new ones using `docker compose rm`, I'm not sure if this the right way of getting it to work.


<img width="1076" alt="Screenshot 2023-04-12 at 2 54 44 PM" src="https://user-images.githubusercontent.com/33514341/231594447-62c8e620-6509-4e93-9806-2bb320efb69b.png">

**Comment by hardikpatil** [comment]
@AbhinavReddy-Dev @shubhamlatkar  I removed all the images and stopped containers and restarted  `docker-compose build` and `docker-compose up` all over again in the root directory. It seems to be working fine @kbuffardi  

![Screenshot 2023-04-12 at 15 45 44](https://user-images.githubusercontent.com/12888989/231602653-0e254403-c640-47b2-bdd3-d77251ab7e87.png)

**Comment by AbhinavReddy-Dev** [comment]
How do we add more packages to the project so that the docker image is up to date? Do we install the packages using docker shell for each of the containers in frontend and backend? This is a naive question but wanted to confirm from others.

**Comment by briswells** [comment]
@AbhinavReddy-Dev I just linked a branch to the issue. Can you test it out using the following steps?

git fetch
git checkout docker-module-issue
docker compose build
docker compose up

//Check everything works

cd backend
npm install mocha --save
cd ../PantryNodeReact
npm install mocha --save

//Add require("mocha") into backend/app.js and  PantryNodeReact/src/App.js

docker compose build
docker compse up

**Comment by briswells** [comment]
> 

The docker compose and dockerfiles "should" be doing that. When we install something locally it updates package.json which during docker compose build is than copied to the container and installed. I think what is happening is the mount in docker compose for node module was doing the opposite of what we intended and overwritting node modules after the install occurred. So instead of saving node modules for the next build, it just stuck whatever ones were in the container when its volumes were created back.

**Comment by hardikpatil** [comment]
> @AbhinavReddy-Dev @hardikpatil I just linked a branch to the issue. Can you test it out using the following steps?
> 
> `git fetch` `git checkout docker-module-issue` `docker compose build` `docker compose up`
> 
> //Check everything works
> 
> `cd backend` `npm install mocha --save` `cd ../PantryNodeReact` `npm install mocha --save`
> 
> //Add require("mocha") into backend/app.js and PantryNodeReact/src/App.js
> 
> `docker compose build` `docker compose up`

though, I see `sequelize` already installed in the `package.json`

![Screenshot 2023-04-12 at 16 53 24](https://user-images.githubusercontent.com/12888989/231610768-634e9689-4e69-4c3f-9f0a-6a08515c86fe.png)

**Comment by hardikpatil** [comment]
> @hardikpatil I'm so confused. I updated the branch, could you pull changes and run the test again?

Did. Also, `node_module` directory is empty, do you think it has something to do with that?

![Screenshot 2023-04-12 at 19 07 38](https://user-images.githubusercontent.com/12888989/231629065-ff8c49bf-9e60-4727-8028-2380838d5478.png)

**Comment by AbhinavReddy-Dev** [comment]
> @AbhinavReddy-Dev The docker compose and dockerfiles "should" be doing that. When we install something locally it updates package.json which during docker compose build is than copied to the container and installed. I think what is happening is the mount in docker compose for node module was doing the opposite of what we intended and overwritting node modules after the install occurred. So instead of saving node modules for the next build, it just stuck whatever ones were in the container when its volumes were created back.

This totally makes sense.

**Comment by briswells** [comment]
> > @hardikpatil I'm so confused. I updated the branch, could you pull changes and run the test again?
> 
> Did. Also, `node_module` directory is empty, do you think it has something to do with that?
> 
> ![Screenshot 2023-04-12 at 19 07 38](https://user-images.githubusercontent.com/12888989/231629065-ff8c49bf-9e60-4727-8028-2380838d5478.png)

Node module folder locally or in a container? Locally I couldn't have touched your local one, it's ignored by both docker and github so I'm not sure how it could be empty. You might need to nuke your volumes and try again. Maybe the old volumes that contained the node modules are auto mounting since they exist even though they are no longer in the docker compose.

**Comment by hardikpatil** [comment]
> Node module folder locally or in a container? Locally I couldn't have touched your local one, it's ignored by both docker and github so I'm not sure how it could be empty. You might need to nuke your volumes and try again. Maybe the old volumes that contained the node modules are auto mounting since they exist even though they are no longer in the docker compose

Interesting, could possible have something to do with Docker volumes. Let me prune and give it another shot again.

**Comment by hardikpatil** [comment]
> > Node module folder locally or in a container? Locally I couldn't have touched your local one, it's ignored by both docker and github so I'm not sure how it could be empty. You might need to nuke your volumes and try again. Maybe the old volumes that contained the node modules are auto mounting since they exist even though they are no longer in the docker compose
> 
> Interesting, could possible have something to do with Docker volumes. Let me prune and give it another shot again.

@briswells nah, still the same. @AbhinavReddy-Dev you getting the same?

**Comment by AbhinavReddy-Dev** [comment]
> 



> > > Node module folder locally or in a container? Locally I couldn't have touched your local one, it's ignored by both docker and github so I'm not sure how it could be empty. You might need to nuke your volumes and try again. Maybe the old volumes that contained the node modules are auto mounting since they exist even though they are no longer in the docker compose
> > 
> > 
> > Interesting, could possible have something to do with Docker volumes. Let me prune and give it another shot again.
> 
> @briswells nah, still the same. @AbhinavReddy-Dev you getting the same?

It works for me though, just finished checking it.

**Comment by briswells** [comment]
Okay so I think I figured out why we're all getting different results. Docker uses a cache to speed up build, so since people built at different times their caches will have different version of the containers. These caches remain even after deleting everything (which makes total sense....). 

When I move to a different computer and run through the testing steps I get a totally different result than on the computer I was using yesterday. The backend works for adding and removing modules but the front doesn't build because react-script is not found. I think we might be getting closer to the solution but I'm still not 100% sure what's going on.

**Comment by shubhamlatkar** [comment]
Hello,
Tried to add the fix for the docker issue.
You can check it at [docker-fix](https://github.com/ChicoState/PantryNode/tree/docker-fix) branch of our repository.
I wonder who tested, reviewed and approved the breaking code at the first place. 🤔

**Comment by briswells** [comment]
> Hello, Tried to add the fix for the docker issue. You can check it at [docker-fix](https://github.com/ChicoState/PantryNode/tree/docker-fix) branch of our repository. I wonder who tested, reviewed and approved the breaking code at the first place. 🤔

Well that's the fun thing. Everyone has been getting different results. It works for some and not for others. Guessing the docker cache is the issue so it depends on when you first build the containers, but I can't say for sure

**Comment by briswells** [comment]
Omg I figured out the root issue. Dockerignore doesn't matter at runtime. So when we build everything is installed correctly but when we run and mount our code it mounts the node_modules even though it's ignored. I'm going to try and put together a fix that only changes this, and all other changes can be created as new issues.

**Comment by shubhamlatkar** [comment]
> Omg I figured out the root issue. Dockerignore doesn't matter at runtime. So when we build everything is installed correctly but when we run and mount our code it mounts the node_modules even though it's ignored. I'm going to try and put together a fix that only changes this, and all other changes can be created as new issues.

Have a look at [docker-fix](https://github.com/ChicoState/PantryNode/tree/docker-fix).

**Comment by briswells** [comment]
> 

I did but there were changes that didn't need to occur, such as changed container names, that effects multiple things and instead of going and changes all those as well it's best just to do a bare minimum fix. I'm going to slim it down to the small possible fix to main.

**Comment by shubhamlatkar** [comment]
> I did but there were changes that didn't need to occur, such as changed container names, that effects multiple things and instead of going and changes all those as well it's best just to do a bare minimum fix. I'm going to slim it down to the small possible fix to main.

Oki all the best 👍🏻

**Comment by ysiddhapura** [comment]
> > Omg I figured out the root issue. Dockerignore doesn't matter at runtime. So when we build everything is installed correctly but when we run and mount our code it mounts the node_modules even though it's ignored. I'm going to try and put together a fix that only changes this, and all other changes can be created as new issues.
> 
> Have a look at [docker-fix](https://github.com/ChicoState/PantryNode/tree/docker-fix).

The docker-fix branch works for me, I suggest we merge this branch to main.

**Comment by jayrevolinskyjr** [comment]
> > > @hardikpatil I'm so confused. I updated the branch, could you pull changes and run the test again?
> > 
> > 
> > Did. Also, `node_module` directory is empty, do you think it has something to do with that?
> > ![Screenshot 2023-04-12 at 19 07 38](https://user-images.githubusercontent.com/12888989/231629065-ff8c49bf-9e60-4727-8028-2380838d5478.png)
> 
> Node module folder locally or in a container? Locally I couldn't have touched your local one, it's ignored by both docker and github so I'm not sure how it could be empty. You might need to nuke your volumes and try again. Maybe the old volumes that contained the node modules are auto mounting since they exist even though they are no longer in the docker compose.

I cloned a fresh copy of the repo, docker builds but fails to run. I get the exact same error as shown here. 

> `git fetch`
> `git checkout docker-module-issue`
> `docker compose build`
> `docker compose up`
>
> //Check everything works
>
> `cd backend`
> `npm install mocha --save`
> `cd ../PantryNodeReact`
> `npm install mocha --save`
>
> //Add require("mocha") into backend/app.js and  PantryNodeReact/src> /App.js
>
> `docker compose build`
> `docker compose up`

I ran a fresh clone again and ran the above commands, it still did not work.

I ran a fresh clone for the third time. I deleted the docker cache, ran docker-compose down, and tried again, it still does not work. 

>I did but there were changes that didn't need to occur, such as changed container names, that effects multiple things and instead of going and changes all those as well it's best just to do a bare minimum fix. I'm going to slim it down to the small possible fix to main.

 @briswells After switching to the docker-fix branch and running from there, it works for me. Is the only problem the container names? I am failing to see what the 'bare minimum fix' is that has not already been done in the docker-fix branch. 

I am moving this to an urgent issue in the project board since this is the mother of all blockers.

**Comment by briswells** [comment]
@jayrevolinskyjr @ysiddhapura @AbhinavReddy-Dev @ysiddhapura @hardikpatil 

Per ysiddhapura I moved backend into a src folder. From there all it took was updating mounts and everything began to work as expected in [docker-module-issue)](https://github.com/ChicoState/PantryNode/tree/docker-module-issue). I had to delete containers, volumes, and images and ```docker rebuild --no-cache``` but after that the changes took effect and we can add/remove npm modules to our hearts content.

When I was talking about bare minimum I was also referring to changes like removing mounts which happened in fix docker. This means we'd to rebuild and rerun docker after every code change vs it being live. With the mounts we'll only need to build the container if modules are added or deleted.

Either way I tested it on 2 computer and it appears to be working as expected.
